File: 347. Top K Frequent Elements/347. Top K Frequent Elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>((a,b) -> a.getValue() - b.getValue());
        Map<Integer, Integer> mp = new HashMap<>();
        for(int n: nums){
            mp.put(n, mp.getOrDefault(n,0)+1);
        }
        for(Map.Entry<Integer,Integer> pair: mp.entrySet()){
            if(pq.size() < k){
                pq.add(pair);
            }
            else if(pair.getValue() > pq.peek().getValue()){
                pq.poll();
                pq.add(pair);
            }
        }
        int[] ans = new int[k];
        for(int i = 0; i < k; ++i){
            ans[i] = pq.poll().getKey();
        }
        return ans;
    }
}

File: 347. Top K Frequent Elements/347. Top K Frequent Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> um;
        for(int& n: nums){
            um[n]++;
        }
        priority_queue<pair<int, int>, vector<pair<int, int>>, MyComp> pq;
        for(auto& p: um){
            pq.push(p);
            if(pq.size() > k)
                pq.pop();
        }
        vector<int> ans;
        while(!pq.empty()) {
            ans.push_back(pq.top().first);
            pq.pop();
        }
        return ans;
    }
private:
    struct MyComp {
        bool operator()(const pair<int, int>& a, const pair<int, int>& b){
            return a.second > b.second;
        }
    };
};